prog/dns_zone/setup_dns_server_vm.rb

```diff
@@ -0,0 +1,137 @@
+    unless (project = Project[project_id])
+      fail "No existing Project"
+    end
+
+    name ||= dns_server.ubid.to_s + SecureRandom.alphanumeric(8).downcase
+
+    DB.transaction do
+      vm_st = Prog::Vm::Nexus.assemble_with_sshable(
+        "ubi",
+        project.id,
+        location: location,
+        name: name,
+        size: vm_size,
+        storage_volumes: [
+          {encrypted: true, size_gib: storage_size_gib}
+        ],
+        boot_image: "ubuntu-jammy",
+        enable_ip4: true
+      )
+
+      Strand.create_with_id(prog: "DnsZone::SetupDnsServerVm", label: "configure_knot", stack: [{subject_id: vm_st.id, dns_server_id: dns_server_id}])
+    end
+  end
+
+  def ds
+    @ds ||= DnsServer[frame["dns_server_id"]]
+  end
+
+  label def configure_knot
+    nap 5 unless vm.strand.label == "wait"
+
+    register_deadline(nil, 15 * 60)
+
+    sshable.cmd <<~SH
+sudo sed -i 's/#DNSStubListener=yes/DNSStubListener=no/' /etc/systemd/resolved.conf
+sudo sed -i ':a;N;$!ba;s/127.0.0.1 localhost\\n\\n#/127.0.0.1 localhost\\n127.0.0.1 #{vm.inhost_name}\\n\\n#/' /etc/hosts
+sudo ln -sf /run/systemd/resolve/resolv.conf /etc/resolv.conf
+sudo apt-get update
+sudo apt-get -y install apt-transport-https ca-certificates wget
+sudo wget -O /usr/share/keyrings/cznic-labs-pkg.gpg https://pkg.labs.nic.cz/gpg
+echo "deb [signed-by=/usr/share/keyrings/cznic-labs-pkg.gpg] https://pkg.labs.nic.cz/knot-dns jammy main" | sudo tee /etc/apt/sources.list.d/cznic-labs-knot-dns.list
+sudo apt-get update
+sudo apt-get -y install knot
+
+echo "KNOTD_ARGS="-C /var/lib/knot/confdb"" | sudo tee -a /etc/default/knot
+    SH
+
+    knot_config = <<-CONF
+server:
+    rundir: "/run/knot"
+    user: "knot:knot"
+    listen: [ "0.0.0.0@53", "::@53" ]
+
+log:
+  - target: "syslog"
+    any: "info"
+
+database:
+    storage: "/var/lib/knot"
+
+acl:
+  - id: "allow_dynamic_updates"
+    address: "127.0.0.1/32"
+    action: "update"
+
+template:
+  - id: "default"
+    storage: "/var/lib/knot"
+    file: "%s.zone"
+    acl: "allow_dynamic_updates"
+    zonefile-sync: "60"
+    zonefile-load: "difference"
+    journal-content: "all"
+
+
+zone:
+  #{ds.dns_zones.map { |dz| "- domain: \"#{dz.name}.\"" }.join("\n  ")}
+    CONF
+
+    sshable.cmd("sudo tee /etc/knot/knot.conf > /dev/null", stdin: knot_config)
+
+    sshable.cmd("sudo systemctl reboot")
+    hop_sync_zones
+  end
+
+  label def sync_zones
+    nap 5 unless sshable.available?
+
+    nap 5 if ds.dns_zones.any?(&:refresh_dns_servers_set?)
+
+    ds.dns_zones.each do |dz|
+      zone_config = <<-CONF
+#{dz.name}.          3600    SOA     ns.#{dz.name}. #{dz.name}. 37 86400 7200 1209600 3600
+#{dz.name}.          3600    NS      elephant.ns.ubicloud.com.
+      CONF
+      sshable.cmd("sudo -u knot tee /var/lib/knot/#{dz.name}.zone > /dev/null", stdin: zone_config)
+    end
+
+    sshable.cmd "sudo systemctl restart knot"
+
+    ds.dns_zones.each(&:purge)
+
+    commands = ds.dns_zones.flat_map do |dz|
```

Comment:
I don't think this is concurrency safe. What if something changes in the dns zone after we read the records?

Reply:
Yeah, it's not concurrency-safe, but I would prefer putting it as a future work. I've already added that to the backlog.

Reply:
For development, I think this is fine.